vm/systemSmartContracts/esdt.go

```diff
@@ -795,6 +805,206 @@ func (e *esdt) controlChanges(args *vmcommon.ContractCallInput) vmcommon.ReturnC
 	return vmcommon.Ok
 }
 
+func (e *esdt) checkArgumentsForSpecialRoleChanges(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```

Comment:
This file grew a lot with this PR. Maybe reorganize code in multiple files? Better: let's split the esdt functionality in a few subcomponents! Otherwise we risk to follow the scProcessor path: a file with a struct that does it all.
As a first step, I would put all check functions in a specialized subcomponent that will be extended by the esdt struct. (using an unexported interface will greatly simplify the testing process). Can be done in another PR, let's discuss about this.

Reply:
this is a contract. it is still smaller than validator.go :D